File: moonball_archiver.rs
use std::fs::{File, OpenOptions};
use std::io::{Read, Write, BufReader, BufWriter, Seek, SeekFrom};
use zstd::stream::{copy_encode, copy_decode};
use brotli;
use lzma;
use serde::{Serialize, Deserialize};
use rayon::prelude::*;
use clap::{App, Arg};
use std::collections::HashMap;
use std::sync::Mutex;
use crypto::aes::{KeySize, cbc_encryptor, cbc_decryptor};
use crypto::buffer::{RefReadBuffer, RefWriteBuffer, BufferResult};
use crypto::symmetriccipher::{Encryptor, Decryptor};
use crypto::sha2::Sha256;
use crypto::hmac::Hmac;
use crypto::mac::Mac;
use eframe::egui;
use ndarray::Array1;
use std::process::Command;
use std::time::{SystemTime, UNIX_EPOCH};
use base32::Alphabet::RFC4648;
use base32;
use rand::Rng;
use totp_lite::totp_custom;
use log::{info, error};

#[derive(Serialize, Deserialize, Debug)]
struct ChunkMetadata {
    file_name: String,
    chunk_id: usize,
    original_size: usize,
    compressed_size: usize,
    compression_algo: String,
    embedding: Option<Vec<f32>>, // Embedding for semantic search
}

#[derive(Serialize, Deserialize, Debug)]
struct ArchiveMetadata {
    chunks: Vec<ChunkMetadata>,
    encryption_enabled: bool,
    requires_2fa: bool,
    secret_key: Option<String>, // 2FA Secret Key
}

pub struct MoonBallArchive {
    metadata: ArchiveMetadata,
    files: Mutex<HashMap<String, Vec<u8>>>, // Cache to store compressed chunks in memory for parallel operations
    config: Config,
}

#[derive(Serialize, Deserialize, Debug)]
struct Config {
    preset: String,
    custom_models: HashMap<String, String>,
    multi_modal_model: String,
    fallback_model: String,
    fallback_provider_settings: FallbackProviderSettings,
    caching: bool,
    model_download_path: String,
    logging_level: String,
    parallel_threads: String,
    compression_level: u8,
    chunk_size: usize,
    error_handling: String,
    auto_update_models: bool,
    scheme: String,
    preset_configs: HashMap<String, PresetConfig>,
    semantic_search: SemanticSearchConfig,
    encryption: EncryptionConfig,
    two_factor_authentication: TwoFactorAuthenticationConfig,
    compression_algorithms: CompressionAlgorithmsConfig,
    logging: LoggingConfig,
}

#[derive(Serialize, Deserialize, Debug)]
struct FallbackProviderSettings {
    ollama: OllamaSettings,
    openai: OpenAISettings,
}

#[derive(Serialize, Deserialize, Debug)]
struct OllamaSettings {
    provider: String,
    model: String,
    contextLength: u32,
    systemMessage: String,
}

#[derive(Serialize, Deserialize, Debug)]
struct OpenAISettings {
    provider: String,
    model: String,
    contextLength: u32,
    apiKey: String,
    apiBase: String,
    systemMessage: String,
}

#[derive(Serialize, Deserialize, Debug)]
struct PresetConfig {
    custom_models: HashMap<String, String>,
    multi_modal: bool,
}

#[derive(Serialize, Deserialize, Debug)]
struct SemanticSearchConfig {
    threshold: f32,
}

#[derive(Serialize, Deserialize, Debug)]
struct EncryptionConfig {
    algorithm: String,
}

#[derive(Serialize, Deserialize, Debug)]
struct TwoFactorAuthenticationConfig {
    enabled: bool,
    issuer: String,
    algorithm: String,
    period: u32,
}

#[derive(Serialize, Deserialize, Debug)]
struct CompressionAlgorithmsConfig {
    brotli: BrotliConfig,
    lzma: LzmaConfig,
    zstd: ZstdConfig,
}

#[derive(Serialize, Deserialize, Debug)]
struct BrotliConfig {
    level: u32,
}

#[derive(Serialize, Deserialize, Debug)]
struct LzmaConfig {
    preset: String,
}

#[derive(Serialize, Deserialize, Debug)]
struct ZstdConfig {
    level: i32,
}

#[derive(Serialize, Deserialize, Debug)]
struct LoggingConfig {
    file_path: String,
    max_size_mb: u32,
    backup_count: u32,
}

impl MoonBallArchive {
    pub fn new() -> Self {
        let config = load_config().expect("Failed to load configuration");
        MoonBallArchive {
            metadata: ArchiveMetadata {
                chunks: Vec::new(),
                encryption_enabled: false,
                requires_2fa: config.two_factor_authentication.enabled,
                secret_key: None, // Initialize with actual secret key
            },
            files: Mutex::new(HashMap::new()),
            config,
        }
    }

    pub fn add_file(&self, file_path: &str) -> Result<(), Box<dyn std::error::Error>> {
        let file = File::open(file_path)?;
        let mut reader = BufReader::new(file);
        let mut buffer = vec![0; self.config.chunk_size];
        let mut chunk_id = 0;

        while let Ok(size) = reader.read(&mut buffer) {
            if size == 0 {
                break;
            }
            self.compress_chunk(&buffer[..size], file_path, chunk_id)?;
            chunk_id += 1;
        }

        Ok(())
    }

    fn compress_chunk(&self, chunk: &[u8], file_path: &str, chunk_id: usize) -> Result<(), Box<dyn std::error::Error>> {
        let algo = self.predict_compression_algo(chunk);
        let compressed_data = match algo.as_str() {
            "brotli" => brotli::CompressorReader::new(chunk, 4096, self.config.compression_algorithms.brotli.level as usize).read_to_end(&mut Vec::new())?,
            "lzma" => lzma::LzmaWriter::new_compressor(Vec::new(), self.config.compression_algorithms.lzma.preset.as_str()).write_all(chunk)?.get_buffer().to_vec(),
            _ => {
                let mut output = Vec::new();
                copy_encode(&mut &chunk[..], &mut output, self.config.compression_algorithms.zstd.level)?;
                output
            }
        };

        if self.metadata.encryption_enabled {
            let encrypted_data = self.encrypt_data(&compressed_data)?;
            compressed_data = encrypted_data;
        }

        let base64_chunk = base64::encode(chunk);
        let output = Command::new("python3")
            .arg("generate_embedding.py")
            .arg(base64_chunk)
            .output()?;
        
        if !output.status.success() {
            return Err(Box::new(std::io::Error::new(
                std::io::ErrorKind::Other,
                format!("Failed to generate embedding: {}", String::from_utf8_lossy(&output.stderr)),
            )));
        }

        let output_str = String::from_utf8(output.stdout)?;
        let embedding: Vec<f32> = serde_json::from_str(&output_str)?;

        let metadata = ChunkMetadata {
            file_name: file_path.to_string(),
            chunk_id,
            original_size: chunk.len(),
            compressed_size: compressed_data.len(),
            compression_algo: algo.clone(),
            embedding: Some(embedding),
        };

        self.metadata.chunks.push(metadata);

        // Save compressed data to memory cache
        let mut files = self.files.lock().unwrap();
        files.insert(format!("{}_{}.mbc", file_path, chunk_id), compressed_data);

        Ok(())
    }

    fn predict_compression_algo(&self, _chunk: &[u8]) -> String {
        // Placeholder ML logic
        "zstd".to_string()
    }

    pub fn save_archive(&self, archive_path: &str) -> Result<(), Box<dyn std::error::Error>> {
        let archive_file = File::create(archive_path)?;
        let mut writer = BufWriter::new(archive_file);

        let metadata_json = serde_json::to_string(&self.metadata)?;

        if self.metadata.encryption_enabled {
            let encrypted_metadata = self.encrypt_data(metadata_json.as_bytes())?;
            writer.write_all(&encrypted_metadata)?;
        } else {
            writer.write_all(metadata_json.as_bytes())?;
        }

        writer.write_all(b"\n")?;

        for (file_name, data) in self.files.lock().unwrap().iter() {
            writer.write_all(data)?;
            writer.write_all(b"\n")?;
        }

        let checksum = format!("{:x}", md5::compute(metadata_json));
        if self.metadata.encryption_enabled {
            let encrypted_checksum = self.encrypt_data(checksum.as_bytes())?;
            writer.write_all(&encrypted_checksum)?;
        } else {
            writer.write_all(checksum.as_bytes())?;
        }

        Ok(())
    }

    pub fn extract(&self, archive_path: &str, output_dir: &str) -> Result<(), Box<dyn std::error::Error>> {
        if self.metadata.requires_2fa {
            let secret_key = self.metadata.secret_key.clone().ok_or("2FA secret key missing")?;
            let current_time = SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs();
            let expected_otp = totp_custom::<Sha256>(&secret_key, current_time, 30, 6, &RFC4648 { padding: false });
            if let Some(provided_otp) = self.ask_for_otp() {
                if provided_otp != expected_otp {
                    return Err(Box::new(std::io::Error::new(
                        std::io::ErrorKind::PermissionDenied,
                        "Invalid OTP",
                    )));
                }
            } else {
                return Err(Box::new(std::io::Error::new(
                    std::io::ErrorKind::PermissionDenied,
                    "OTP required for extraction",
                )));
            }
        }

        let archive_file = File::open(archive_path)?;
        let mut reader = BufReader::new(archive_file);
        let mut metadata_str = String::new();
        reader.read_line(&mut metadata_str)?;

        if self.metadata.encryption_enabled {
            let decrypted_metadata = self.decrypt_data(metadata_str.as_bytes())?;
            let metadata: ArchiveMetadata = serde_json::from_slice(&decrypted_metadata)?;
        } else {
            let metadata: ArchiveMetadata = serde_json::from_str(&metadata_str)?;
        }

        for (index, chunk_metadata) in metadata.chunks.iter().enumerate() {
            let mut compressed_data = vec![0; chunk_metadata.compressed_size];
            reader.read_exact(&mut compressed_data)?;

            if self.metadata.encryption_enabled {
                let decrypted_chunk = self.decrypt_data(&compressed_data)?;
                compressed_data = decrypted_chunk;
            }

            let output_path = std::path::Path::new(output_dir).join(&chunk_metadata.file_name);
            let mut output_file = File::create(&output_path)?;

            match chunk_metadata.compression_algo.as_str() {
                "brotli" => {
                    let mut decompressor = brotli::Decompressor::new(&compressed_data[..], 4096);
                    std::io::copy(&mut decompressor, &mut output_file)?;
                }
                "lzma" => {
                    let mut decompressor = lzma::LzmaReader::new_decompressor(&compressed_data[..])?;
                    std::io::copy(&mut decompressor, &mut output_file)?;
                }
                _ => {
                    copy_decode(&mut &compressed_data[..], &mut output_file)?;
                }
            }
        }

        Ok(())
    }

    fn encrypt_data(&self, data: &[u8]) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
        let salt = rand::thread_rng().gen::<[u8; 16]>();
        let kdf = Hmac::<Sha256>::new_from_slice(b"moonball").unwrap();
        let mut mac = kdf.clone();
        mac.update(&salt);
        let key = mac.finalize().into_bytes();

        let iv = rand::thread_rng().gen::<[u8; 16]>();
        let cipher = cbc_encryptor(KeySize::KeySize256, &key[..], &iv[..], crypto::blockmodes::PkcsPadding);
        let mut final_result = Vec::<u8>::new();
        let mut read_buffer = RefReadBuffer::new(data);
        let mut buffer = [0; 4096];
        let mut write_buffer = RefWriteBuffer::new(&mut buffer);

        loop {
            let result = cipher.encrypt(&mut read_buffer, &mut write_buffer, true)?;
            final_result.extend(write_buffer.take_read_buffer().take_remaining());
            match result {
                BufferResult::BufferUnderflow => break,
                BufferResult::BufferOverflow => {}
            }
        }

        Ok([salt.to_vec(), iv.to_vec(), final_result].concat())
    }

    fn decrypt_data(&self, data: &[u8]) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
        let salt = &data[..16];
        let iv = &data[16..32];
        let kdf = Hmac::<Sha256>::new_from_slice(b"moonball").unwrap();
        let mut mac = kdf.clone();
        mac.update(salt);
        let key = mac.finalize().into_bytes();

        let cipher = cbc_decryptor(KeySize::KeySize256, &key[..], iv, crypto::blockmodes::PkcsPadding);
        let mut final_result = Vec::<u8>::new();
        let mut read_buffer = RefReadBuffer::new(&data[32..]);
        let mut buffer = [0; 4096];
        let mut write_buffer = RefWriteBuffer::new(&mut buffer);

        loop {
            let result = cipher.decrypt(&mut read_buffer, &mut write_buffer, true)?;
            final_result.extend(write_buffer.take_read_buffer().take_remaining());
            match result {
                BufferResult::BufferUnderflow => break,
                BufferResult::BufferOverflow => {}
            }
        }

        Ok(final_result)
    }

    fn ask_for_otp(&self) -> Option<String> {
        println!("Enter 2FA OTP: ");
        let mut input = String::new();
        std::io::stdin().read_line(&mut input).ok()?;
        Some(input.trim().to_string())
    }
}

fn load_config() -> Result<Config, Box<dyn std::error::Error>> {
    let config_path = "config.yml";
    let file = File::open(config_path)?;
    let reader = BufReader::new(file);
    let config: Config = serde_yaml::from_reader(reader)?;

    Ok(config)
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    env_logger::Builder::from_default_env()
        .filter_level(log::LevelFilter::Info)
        .init();

    let matches = App::new("MoonBall Archiver")
        .version("1.0")
        .author("Your Name <your.email@example.com>")
        .about("A Rust implementation of MoonBall Archiver")
        .arg(
            Arg::with_name("add")
                .short('a')
                .long("add")
                .value_name("FILES/DIRECTORIES")
                .help("Files or directories to add to the archive")
                .multiple(true)
                .takes_value(true),
        )
        .arg(
            Arg::with_name("extract")
                .short('e')
                .long("extract")
                .value_name("ARCHIVE")
                .help("Archive to extract")
                .takes_value(true),
        )
        .arg(
            Arg::with_name("output")
                .short('o')
                .long("output")
                .value_name("DIR/FILENAME")
                .help("Output directory or archive name")
                .takes_value(true)
                .required_unless_one(&["add", "gui"]),
        )
        .arg(
            Arg::with_name("scheme")
                .short('s')
                .long("scheme")
                .possible_values(&["fast", "balanced", "max"])
                .default_value("balanced")
                .help("Compression scheme"),
        )
        .arg(
            Arg::with_name("extension")
                .short('x')
                .long("extension")
                .possible_values(&["mnbl", "🌕"])
                .default_value("mnbl")
                .help("File extension for the archive"),
        )
        .arg(
            Arg::with_name("gui")
                .short('g')
                .long("gui")
                .help("Launch graphical user interface"),
        )
        .arg(
            Arg::with_name("search")
                .short('S')
                .long("search")
                .value_name("QUERY")
                .help("Semantic search query")
                .takes_value(true),
        )
        .get_matches();

    if matches.is_present("gui") {
        // Launch GUI here (not implemented yet)
        println!("GUI not implemented yet.");
    } else if let Some(files) = matches.values_of("add") {
        let output_path = matches.value_of("output").unwrap();
        let extension = matches.value_of("extension").unwrap();

        let mut archive = MoonBallArchive::new();
        for file in files {
            if std::path::Path::new(file).is_dir() {
                // Add directory logic here (not implemented yet)
                println!("Adding directory: {}", file);
            } else {
                archive.add_file(file)?;
            }
        }

        let final_output_path = format!("{}.{}", output_path, extension);
        archive.save_archive(&final_output_path)?;
        info!("Archive saved as {}", final_output_path);
    } else if let Some(archive_path) = matches.value_of("extract") {
        let output_dir = matches.value_of("output").unwrap();
        let mut archive = MoonBallArchive::new(); // Initialize with metadata
        archive.extract(archive_path, output_dir)?;
        info!("Files extracted to {}", output_dir);
    } else if let Some(query) = matches.value_of("search") {
        let mut archive = MoonBallArchive::new(); // Initialize with metadata
        let results = archive.semantic_search(query); // Implement semantic search in Rust (not implemented yet)
        println!("Search results: {:?}", results);
    }

    Ok(())
}

File: moonball_archiver.py
import os
import zstandard as zstd
import brotli
import lzma
import hashlib
from tqdm import tqdm
from concurrent.futures import ThreadPoolExecutor
from sklearn.ensemble import RandomForestClassifier
import numpy as np
from transformers import pipeline
import shutil
import tkinter as tk
from tkinter import filedialog, messagebox, simpledialog
from PIL import Image, ImageTk
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
import base64
import json
import argparse
import pyotp

class MoonBallArchive:
    def __init__(self, encryption_key=None):
        self.files = []
        self.index = []
        self.encryption_key = encryption_key
        self.ml_model = RandomForestClassifier()
        self.embedding_model = pipeline('feature-extraction', model='distilbert-base-uncased')
        self.load_config()

    def load_config(self):
        with open('config.yml', 'r') as config_file:
            import yaml
            self.config = yaml.safe_load(config_file)

    def add_file(self, file_path, chunk_size=5242880):  # Default chunk size: 5MB
        with open(file_path, 'rb') as f:
            chunk_id = 0
            while chunk := f.read(chunk_size):
                self.compress_chunk(chunk, file_path, chunk_id)
                chunk_id += 1

    def compress_chunk(self, chunk, file_path, chunk_id):
        algo = self.predict_compression_algo(file_path, chunk)
        compressed_data = self.compress_data(chunk, algo)

        if self.encryption_key:
            compressed_data = self.encrypt_data(compressed_data)

        embedding = self.generate_embedding(chunk)

        chunk_entry = {
            'file_name': file_path,
            'chunk_id': chunk_id,
            'original_size': len(chunk),
            'compressed_size': len(compressed_data),
            'compression_algo': algo,
            'data': compressed_data,
            'embedding': embedding
        }

        self.files.append(chunk_entry)
        self.index.append({
            'file_name': file_path,
            'chunk_id': chunk_id,
            'original_size': len(chunk),
            'compressed_size': len(compressed_data),
            'compression_algo': algo,
            'embedding': embedding
        })

    def predict_compression_algo(self, file_path, chunk):
        # Placeholder ML logic
        features = np.array([[len(chunk), self.calculate_entropy(chunk)]])
        algo_index = self.ml_model.predict(features)[0]
        algo_map = {0: 'brotli', 1: 'lzma', 2: 'zstd'}
        return algo_map.get(algo_index, 'zstd')

    def calculate_entropy(self, data):
        if len(data) == 0:
            return 0
        byte_counts = [0] * 256
        for byte in data:
            byte_counts[byte] += 1
        entropy = 0
        for count in byte_counts:
            if count > 0:
                p = count / len(data)
                entropy -= p * math.log2(p)
        return entropy

    def generate_embedding(self, chunk):
        base64_chunk = base64.b64encode(chunk).decode('utf-8')
        output = Command('python3').args(['generate_embedding.py', base64_chunk]).output()
        if not output.status.success():
            raise Exception(f"Failed to generate embedding: {output.stderr}")
        return json.loads(output.stdout)

    def add_files_parallel(self, file_paths):
        with ThreadPoolExecutor() as executor:
            list(tqdm(executor.map(self.add_file, file_paths), total=len(file_paths), desc='Indexing and Compressing Files'))

    def add_directory(self, dir_path):
        for root, _, files in os.walk(dir_path):
            for file in files:
                self.add_file(os.path.join(root, file))

    def save(self, archive_path):
        if not (archive_path.endswith(".mnbl") or archive_path.endswith("\U0001F315")):
            raise ValueError("Invalid file extension. Please use '.mnbl' or '.🌕' for the archive.")

        with open(archive_path, 'wb') as archive:
            index_data = str(self.index).encode('utf-8')
            if self.encryption_key:
                index_data = self.encrypt_data(index_data)
            archive.write(index_data)
            archive.write(b'\n')

            for file in tqdm(self.files, desc='Writing chunks to archive'):
                archive.write(file['data'])
                archive.write(b'\n')

            checksum = hashlib.sha256(index_data).hexdigest().encode('utf-8')
            if self.encryption_key:
                checksum = self.encrypt_data(checksum)
            archive.write(checksum)

    def extract(self, archive_path, output_dir):
        if not (archive_path.endswith(".mnbl") or archive_path.endswith("\U0001F315")):
            raise ValueError("Invalid file extension. Please use '.mnbl' or '.🌕' for the archive.")

        with open(archive_path, 'rb') as archive:
            lines = archive.readlines()

            index_data = lines[0].strip()
            if self.encryption_key:
                index_data = self.decrypt_data(index_data)
            index_data = eval(index_data.decode('utf-8'))

            for file_meta in tqdm(index_data, desc='Extracting chunks'):
                compressed_data = lines[file_meta['chunk_id'] + 1].strip()

                if self.encryption_key:
                    compressed_data = self.decrypt_data(compressed_data)

                decompressed_data = self.decompress_data(compressed_data, file_meta['compression_algo'])

                output_path = os.path.join(output_dir, os.path.basename(file_meta['file_name']))
                mode = 'wb' if file_meta['chunk_id'] == 0 else 'ab'
                with open(output_path, mode) as out_file:
                    out_file.write(decompressed_data)

    def semantic_search(self, query):
        query_embedding = self.embedding_model([query])[0][0]
        results = []
        for file_meta in self.index:
            similarity = np.dot(query_embedding, file_meta['embedding'])
            if similarity > self.config['semantic_search']['threshold']:
                results.append(file_meta['file_name'])
        return results

    def encrypt_data(self, data):
        salt = os.urandom(16)
        kdf = PBKDF2HMAC(
            algorithm=hashlib.sha256(),
            length=32,
            salt=salt,
            iterations=100000,
            backend=default_backend()
        )
        key = base64.urlsafe_b64encode(kdf.derive(self.encryption_key.encode('utf-8')))
        iv = os.urandom(16)
        cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
        encryptor = cipher.encryptor()
        padder = self.padding.PKCS7(algorithms.AES.block_size).padder()
        padded_data = padder.update(data) + padder.finalize()
        encrypted_data = encryptor.update(padded_data) + encryptor.finalize()
        return salt + iv + encrypted_data

    def decrypt_data(self, data):
        salt = data[:16]
        iv = data[16:32]
        kdf = PBKDF2HMAC(
            algorithm=hashlib.sha256(),
            length=32,
            salt=salt,
            iterations=100000,
            backend=default_backend()
        )
        key = base64.urlsafe_b64encode(kdf.derive(self.encryption_key.encode('utf-8')))
        cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
        decryptor = cipher.decryptor()
        decrypted_padded_data = decryptor.update(data[32:]) + decryptor.finalize()
        unpadder = self.padding.PKCS7(algorithms.AES.block_size).unpadder()
        decrypted_data = unpadder.update(decrypted_padded_data) + unpadder.finalize()
        return decrypted_data

    def compress_data(self, data, algo):
        if algo == 'brotli':
            compressed_data = brotli.compress(data)
        elif algo == 'lzma':
            compressed_data = lzma.compress(data)
        else:
            cctx = zstd.ZstdCompressor(level=9)  # Adjust level as needed
            compressed_data = cctx.compress(data)
        return compressed_data

    def decompress_data(self, data, algo):
        if algo == 'brotli':
            decompressed_data = brotli.decompress(data)
        elif algo == 'lzma':
            decompressed_data = lzma.decompress(data)
        else:
            dctx = zstd.ZstdDecompressor()
            decompressed_data = dctx.decompress(data)
        return decompressed_data

    def clean_up_temporary_files(self):
        for file_meta in self.files:
            chunk_path = f"{file_meta['file_name']}_{file_meta['chunk_id']}.mbc"
            if os.path.exists(chunk_path):
                os.remove(chunk_path)

def launch_gui():
    def add_files():
        files = filedialog.askopenfilenames()
        if files:
            for file in files:
                file_listbox.insert(tk.END, file)

    def add_directory():
        dir_path = filedialog.askdirectory()
        if dir_path:
            file_listbox.insert(tk.END, dir_path)

    def compress_files():
        files = file_listbox.get(0, tk.END)
        if not files:
            messagebox.showerror('Error', 'No files or directories selected for compression.')
            return
        output_path = filedialog.asksaveasfilename(defaultextension='.mnbl', filetypes=[('MoonBall Archive', '*.mnbl'), ('MoonBall Archive Emoji', '🌕')])
        if not output_path:
            return
        encryption_key = simpledialog.askstring("Encryption Key", "Enter an encryption key (or leave blank for no encryption):", show='*')
        archive = MoonBallArchive(encryption_key=encryption_key)
        for item in files:
            if os.path.isdir(item):
                archive.add_directory(item)
            else:
                archive.add_file(item)
        archive.save(output_path)
        messagebox.showinfo('Success', f'Archive saved as {output_path}')

    def extract_archive():
        archive_path = filedialog.askopenfilename(filetypes=[('MoonBall Archive', '*.mnbl'), ('MoonBall Archive Emoji', '🌕')])
        if not archive_path:
            return
        output_dir = filedialog.askdirectory()
        if not output_dir:
            return
        encryption_key = simpledialog.askstring("Encryption Key", "Enter the encryption key (if set):", show='*')
        archive = MoonBallArchive(encryption_key=encryption_key)
        secret_key = 'your_secret_key_here'  # Replace with actual secret key
        otp = simpledialog.askstring("2FA OTP", "Enter the 2FA OTP (if set):", show='*')
        if verify_otp(secret_key, otp):
            archive.extract(archive_path, output_dir)
            messagebox.showinfo('Success', f'Files extracted to {output_dir}')
        else:
            messagebox.showerror('Error', 'Invalid OTP.')

    def semantic_search():
        query = simpledialog.askstring("Semantic Search", "Enter your search query:")
        if query:
            archive = MoonBallArchive()
            results = archive.semantic_search(query)
            result_str = "\n".join(results) if results else "No files found matching the query."
            messagebox.showinfo('Search Results', result_str)

    root = tk.Tk()
    root.title('MoonBall Archiver')
    root.geometry('600x500')

    try:
        moonball_logo = Image.open(os.path.join('assets', 'moonball_file.ico'))
        moonball_logo = moonball_logo.resize((50, 50), Image.LANCZOS)
        logo_img = ImageTk.PhotoImage(moonball_logo)
        logo_label = tk.Label(root, image=logo_img)
        logo_label.image = logo_img
        logo_label.pack(pady=10)
    except FileNotFoundError:
        print('assets/moonball_icon.png not found. Skipping logo display.')

    add_button = tk.Button(root, text='Add Files', command=add_files)
    add_button.pack(pady=5)

    add_dir_button = tk.Button(root, text='Add Directory', command=add_directory)
    add_dir_button.pack(pady=5)

    file_listbox = tk.Listbox(root, selectmode=tk.MULTIPLE, width=60, height=10)
    file_listbox.pack(pady=5)

    compress_button = tk.Button(root, text='Compress to Archive', command=compress_files)
    compress_button.pack(pady=5)

    extract_button = tk.Button(root, text='Extract Archive', command=extract_archive)
    extract_button.pack(pady=5)

    search_button = tk.Button(root, text='Semantic Search', command=semantic_search)
    search_button.pack(pady=5)

    root.mainloop()

def verify_otp(secret_key, provided_otp):
    totp = pyotp.TOTP(secret_key, digest="sha1", digits=6, interval=30)
    return totp.verify(provided_otp)

def main():
    import argparse
    parser = argparse.ArgumentParser(description='MoonBall Archiver')
    parser.add_argument('--add', type=str, nargs='+', help='Files or directories to add to the archive')
    parser.add_argument('--extract', type=str, help='Archive to extract')
    parser.add_argument('--output', type=str, help='Output directory or archive name')
    parser.add_argument('--scheme', choices=['fast', 'balanced', 'max'], default='balanced', help='Compression scheme')
    parser.add_argument('--extension', choices=['mnbl', '🌕'], default='mnbl', help='File extension for the archive')
    parser.add_argument('--gui', action='store_true', help='Launch graphical user interface')
    parser.add_argument('--search', type=str, help='Semantic search query')
    args = parser.parse_args()

    if args.gui:
        launch_gui()
    else:
        encryption_key = simpledialog.askstring("Encryption Key", "Enter an encryption key (or leave blank for no encryption):", show='*') if args.add or args.extract else None
        archive = MoonBallArchive(encryption_key=encryption_key)

        if args.add:
            for item in args.add:
                if os.path.isdir(item):
                    archive.add_directory(item)
                else:
                    archive.add_file(item)
            archive_name = args.output if args.output else f'archive.{args.extension}'
            archive.save(archive_name)
            print(f'Archive saved as {archive_name}')
        elif args.extract:
            if not args.extract or not args.output:
                print('Please provide both the archive to extract and the output directory.')
                return
            os.makedirs(args.output, exist_ok=True)
            secret_key = 'your_secret_key_here'  # Replace with actual secret key
            otp = simpledialog.askstring("2FA OTP", "Enter the 2FA OTP (if set):", show='*')
            if verify_otp(secret_key, otp):
                archive.extract(args.extract, args.output)
                print(f'Files extracted to {args.output}')
            else:
                print("Invalid OTP.")
        elif args.search:
            results = archive.semantic_search(args.search)
            if results:
                print(f"Files matching the query: {', '.join(results)}")
            else:
                print("No files found matching the query.")

if __name__ == '__main__':
    main()